### DISCUSSION TITLE
How does spacy-transformers deal with textcat_multilabel for long documents?

### DISCUSSION THREAD

**Discussion by vsocrates** [explosion/spaCy#12574] [Help: Coding & Implementations] [answered]
I've had a lot of trouble finding this information. How does spacy-transformers handle text classification tasks for long documents? 

By default, I know it uses the overlapping strided spans, but this only seems to make sense for NER, but not as much for  text classification tasks. Thanks!

**Comment by rmitsch** [comment]
Hi @vsocrates!

> How does spacy-transformers handle text classification tasks for long documents?

Document length doesn't influence `spacy-transformers` handles text classification. We do however recommend to split up very long documents into smaller ones to avoid excessive memory consumption - for more information on that see [here](https://github.com/explosion/spaCy/discussions/8600#discussioncomment-969204) and [here](https://github.com/explosion/spaCy/discussions/12109).

> By default, I know it uses the overlapping strided spans, but this only seems to make sense for NER, but not as much for text classification tasks.

Why do you think that overlapping strided spans don't make much sense for text classification? Overlapping spans help with providing more context to the transformer model, which in turn results in richer word representations for downstream components.

**Comment by vsocrates** [comment]
> Why do you think that overlapping strided spans don't make much sense for text classification? Overlapping spans help with providing more context to the transformer model, which in turn results in richer word representations for downstream components.

For instance, if our task is to classify document topics (e.g. science, politics, etc.) and only a couple sentences/words indicate a _politics_ topic, then with overlapping strided spans, there may be a text chunk that will have a _politics_ topic label, but no text indicating this topic. This situation is the one I'm concerned about. 

Would NER make more sense for this setting then?

**Reply by rmitsch** [reply]
To clarify - you want to do a multi-label classification of documents, correct? I'm a bit confused because you mention a text "chunk" that might have a politics label.

**Reply by vsocrates** [reply]
Yes, sorry, I should have been clearer. By "text chunk", I mean one of the strided overlapping spans. 

In the multi-label classification setting, if one of these spans is labeled as positive for the _politics_ label, but the sentence that indicates this label isn't contained within the span, will this not impact model performance?

**Reply by rmitsch** [reply]
Alright, there seems to be a misunderstanding regarding the output for individual spans/windows. The transformer only provides word-level tensors, not `textcat` (or any other) predictions. These tensors are later used to classify the document.

**Reply by vsocrates** [reply]
> The transformer only provides word-level tensors, not textcat (or any other) predictions. These tensors are later used to classify the document.


Sorry, I assumed like in the classic BERT architecture, the vector for the CLS tag is used for document classification. If this isn't the case, then how are the word-level vectors pooled to perform prediction? Would this mean that the spacy pipeline is able to handle arbitrary lengths of documents (e.g. all word-level vectors in a sequence are pooled, despite the 512 dim limitation in BERT)? 

Would we not have issues due to incorrect position embeddings if we're pooling across overlapping spans? 

I apologize, I thought I'd had a better understanding of the internals of spacy than I actually did. Thank you for the help so far!

**Reply by adrianeboyd** [reply]
The pooling and the spans are specified in the config, e.g. an excerpt from `en_core_web_trf/config.cfg`:

```ini
[components.tagger.model.tok2vec]
@architectures = "spacy-transformers.TransformerListener.v1"
grad_factor = 1.0
upstream = "transformer"
# how the wordpiece embeddings are combined for each token for use with the token-level tagger model
# (where each token may also have embeddings from more than one span)
pooling = {"@layers":"reduce_mean.v1"}

[components.transformer]
factory = "transformer"
max_batch_items = 4096
set_extra_annotations = {"@annotation_setters":"spacy-transformers.null_annotation_setter.v1"}

[components.transformer.model]
@architectures = "spacy-transformers.TransformerModel.v3"
name = "roberta-base"
mixed_precision = false

# how a doc is converted into spans
[components.transformer.model.get_spans]
@span_getters = "spacy-transformers.strided_spans.v1"
window = 128
stride = 96
```

More details: https://spacy.io/api/architectures#transformers

**Reply by vsocrates** [reply]
Apologies for the late reply, got pulled away on a few other tasks! 

The pooling to generate word-level tensors makes sense, but I'm then confused how the `textcat` works. Looking at the config below, it looks like the pipeline uses a `TextCatEnsemble` component for prediction. It says this is a BOW model on top of a `tok2vec` neural network, but I don't understand how the word-level tensors are used in the BOW model. 

```
[components.textcat_multilabel]
factory = "textcat_multilabel"
scorer = {"@scorers":"spacy.textcat_multilabel_scorer.v1"}
threshold = 0.5

[components.textcat_multilabel.model]
@architectures = "spacy.TextCatEnsemble.v2"
nO = null

[components.textcat_multilabel.model.linear_model]
@architectures = "spacy.TextCatBOW.v2"
exclusive_classes = false
ngram_size = 1
no_output_layer = false
nO = null

[components.textcat_multilabel.model.tok2vec]
@architectures = "spacy-transformers.TransformerListener.v1"
grad_factor = 1.0
pooling = {"@layers":"reduce_mean.v1"}
upstream = "*"
```

**Reply by danieldk** [reply]
> It says this is a BOW model on top of a tok2vec neural network

It is a (BOW model) _and_ (a neural network on top of tok2vec) (parentheses for disambiguation). The neural network uses a single-headed attention mechanism over the output of the tok2vec pipe (in this case the transformer). The BOW model is applied separately to the document tokens. The model ensemble model combines the predictions of the neural net and the BOW model using a learned output layer.

I hope that clarifies things!

**Reply by vsocrates** [reply]
Hi, yes that makes sense. 

Last question, for the neural network on top of the tok2vec pipe, what is the input size? This seems like it'll determine how big the document can be for **textcat** prediction. _<= This may be an assumption, please clarify if this isn't the case!_ 

Thanks!

**Reply by danieldk** [reply, marked as answer]
Since it applies attention over all the hidden representations in the document, there is no theoretical limit to the input/doc size. Of course there are practical limits (e.g., you'll want to adjust the batch size to the amount of GPU memory available if you run on the GPU).
